### PR TITLE
[WIP] ConstantDynamic: Adds static and dynamic verification for ConDy

### PR DESCRIPTION
**Pull request by taliamccormick**
**WIP PR for ConDy Verification**

This will add static/runtime verification for ConstantDynamic. It checks that:
- ConDy uses occur in classfiles with version Java11 and up
- Validation that `ldc` and `ldc_w` bytecodes receive ConstantDynamic entries which do not return a long (`J`) or double (`D`)
- Validation that `ldc2_w` bytecodes receive ConstantDynamic entries which return a long or double

Resolves https://github.com/eclipse/openj9/issues/1278
Depends on issue https://github.com/eclipse/openj9/issues/1271 with corresponding PRs https://github.com/eclipse/openj9/pull/1426 and https://github.com/eclipse/openj9/pull/1533/files

@fengxue-IS

### CODE REVIEW

File: runtime/bcverify/vrfyhelp.c [hunk 1]
```diff
@@ -417,6 +417,15 @@ pushLdcType(J9ROMClass * romClass, UDATA index, UDATA * stackTop)
 		case J9CPTYPE_METHODHANDLE:
 			PUSH(BCV_JAVA_LANG_INVOKE_METHODHANDLE_INDEX << BCV_CLASS_INDEX_SHIFT);
 			break;
+		case J9CPTYPE_CONSTANT_DYNAMIC:
+			J9ROMConstantDynamicRef* romConstantDynamicRef = (J9ROMConstantDynamicRef *)(J9_ROM_CP_FROM_ROM_CLASS(romClass) + index);
+			J9UTF8 *nameAndSignature = J9ROMCONSTANTDYNAMICREF_NAMEANDSIGNATURE(romConstantDynamicRef);
+			U_8* signature = J9UTF8_DATA(nameAndSignature);
+
+			/* The return value of a method follows the first ')' listed in its signature */
+			while (*signature++ != ')');
```
Comment (fengxue-IS) [inline runtime/bcverify/vrfyhelp.c]: read the condy spec 4.4.13, it seem to suggest that the signature referenced in a ConstantDynamic entry is a fieldDescriptor, so there will be no `()` in the string, it should only contain the return type.

I'll read the spec details to confirm this, may need to add more checks in static verify to consider cases such as double array signature `[D`

Comment (taliamccormick) [reply runtime/bcverify/vrfyhelp.c]: Right, thanks. Looks like I could directly give the value of `J9UTF8_DATA(nameAndSignature)` then, and `pushType` will still handle the rest. 

What situations istatic verify would/would not accept cases such as the double array signature `[[D`? From what I understand:
- Reference types are any non-primitive (eg. objects, arrays)
- Constant dynamic does not place any restrictions on its return value (4.10.1.3)
> dynamic(ConstantName, FieldDescriptor) for a CONSTANT_Dynamic_info, where ConstantName and FieldDescriptor correspond to the name and field descriptor referenced by the name_and_type_index item of the CONSTANT_Dynamic_info structure (the bootstrap_method_attr_index item is irrelevant to verification);

&

> loadableConstantType(CP, Type) :-
CP = dynamic(ConstantName, FieldDescriptor),
parseFieldDescriptor(FieldDescriptor, Type).
- `ldc` & `ldc_w` bytecodes  would still accept the reference types (including double array) and primitives which are not long/double
- `ldc2_w` bytecode would still only accept long/double - but not rerfences, such as `[[D`



---

Quick reread of the section you pointed out gives:

4.4.13 gives:
> 4.4.13 The CONSTANT_Dynamic_info Structure
>  CONSTANT_Dynamic_info {
&nbsp;&nbsp;&nbsp; u1 tag;
&nbsp;&nbsp;   u2 bootstrap_method_attr_index;
&nbsp;&nbsp;   u2 name_and_type_index;
}
...
The value of the name_and_type_index item must be a valid index into the constant_pool table. The constant_pool entry at that index must be a CONSTANT_NameAndType_info structure (4.4.6) representing a name and field descriptor (4.3.2).

4.4.6 gives:
> CONSTANT_NameAndType_info {
&nbsp;&nbsp;&nbsp;    u1 tag;
&nbsp;&nbsp;&nbsp;    u2 name_index;
&nbsp;&nbsp;&nbsp;    u2 descriptor_index;
}
...
The value of the descriptor_index item must be a valid index into the constant_pool table. The constant_pool entry at that index must be a CONSTANT_Utf8_info structure (§4.4.7) representing a valid field descriptor or method descriptor (§4.3.2, §4.3.3).

4.3.2 gives the grammar for a field descriptor as:
> FieldDescriptor:
&nbsp;&nbsp;&nbsp;FieldType
FieldType:
&nbsp;&nbsp;&nbsp;BaseType   |  ObjectType  |  ArrayType
BaseType:
&nbsp;&nbsp;&nbsp; (one of) B C D F I J S Z
ObjectType:
&nbsp;&nbsp;&nbsp;L ClassName ;
ArrayType:
&nbsp;&nbsp;&nbsp;[ ComponentType
ComponentType:
&nbsp;&nbsp;&nbsp;FieldType
(http://cr.openjdk.java.net/~dlsmith/constant-dynamic.html)

Comment (taliamccormick) [reply runtime/bcverify/vrfyhelp.c]: Woops, you're right - sounds like I can put directly into `pushType`

I'm not seeing any restrictions on the type of the signature - references (including cases like `[[D`) are valid to be passed to `ldc` or `ldc_w`

Comment (fengxue-IS) [reply runtime/bcverify/vrfyhelp.c]: your current check is checking at the last `char` of the fieldDescriptor, which would mix up `[[D` and `D` we would have to either check the first char to be `D` or `J`.
the reverse would apply to `ldc` and `ldc_w`

File: runtime/bcverify/staticverify.c [hunk 2]
```diff
@@ -231,6 +233,10 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
 			info = &(classfile->constantPool[index]);
 			tag = (UDATA) info->tag;
 
+			if (CFR_CONSTANT_Dynamic == tag) {
+				constantDynamicNameAndSignature = &classFile->constantPool[classFile->constantPool[info].slot2];
```
Comment (fengxue-IS) [inline runtime/bcverify/staticverify.c]: `&classFile->constantPool[classFile->constantPool[info->slot2].slot2]` missing `->slot2`

File: runtime/bcverify/staticverify.c [hunk 3]
```diff
@@ -257,8 +271,18 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
...
+			}
 			if (!((tag == CFR_CONSTANT_Double)
-				  || (tag == CFR_CONSTANT_Long))) {
+				  || (tag == CFR_CONSTANT_Long)
+				  || ((tag == CFR_CONSTANT_Dynamic)
+						  && (((flags & BCT_MajorClassFileVersionMask) >= BCT_Java11MajorVersionShifted)
+									|| ((flags & BCT_MajorClassFileVersionMask) == 0))
+						  && ('D' != constantDynamicNameAndSignature->bytes[constantDynamicNameAndSignature->slot1 - 1])
+						  && ('J' != constantDynamicNameAndSignature->bytes[constantDynamicNameAndSignature->slot1 - 1]))
```
Comment (fengxue-IS) [inline runtime/bcverify/staticverify.c]: should be `'D' ==` & `'J' ==`

File: runtime/bcverify/staticverify.c [hunk 4]
```diff
@@ -239,7 +245,13 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
 									|| ((flags & BCT_MajorClassFileVersionMask) == 0)))
 				  || (((tag == CFR_CONSTANT_MethodType) || (tag == CFR_CONSTANT_MethodHandle))
 						  && (((flags & BCT_MajorClassFileVersionMask) >= BCT_Java7MajorVersionShifted)
-									|| ((flags & BCT_MajorClassFileVersionMask) == 0)))	)) {
+									|| ((flags & BCT_MajorClassFileVersionMask) == 0)))
+				  || ((tag == CFR_CONSTANT_Dynamic)
+						  && (((flags & BCT_MajorClassFileVersionMask) >= BCT_Java11MajorVersionShifted)
+									|| ((flags & BCT_MajorClassFileVersionMask) == 0))
+						  && ('D' != constantDynamicNameAndSignature->bytes[constantDynamicNameAndSignature->slot1 - 1])
+						  && ('J' != constantDynamicNameAndSignature->bytes[constantDynamicNameAndSignature->slot1 - 1]))
```
[the remark this answers is not in the document]
Comment (taliamccormick) [reply runtime/bcverify/staticverify.c]: Agreed, updating

File: runtime/bcverify/vrfyhelp.c (hunk 1 above)
Comment (taliamccormick) [reply runtime/bcverify/vrfyhelp.c]: Yes, but if it's a a fieldDescriptor, then can't we directly check the first index instead?
ie. it's either a `BaseType`, `ObjectType`, or `ArrayType`
If a `BaseType`, then one of `B C D F I J S Z` will be the first char; if an `ObjectType`, then `L` will be the first char; and if `ArrayType`, then `[` will be the first char

**Comment by taliamccormick** [comment]
Updated
```
    - Copyright dates
    - Refactor ldc for more informative error messages
    - Remove definition of 'J9CPTYPE_CONSTANT_DYNAMIC ' value (defined in
    other PR)
    - Update description comments on pushLdcType(...)
    - Update pushLdcType to expect a constant dynamic to provide a field
    descriptor instead of a method signature
    - Rename variable constantDynamicNameAndSignature to
    constantDynamicSignature
```
and
```
    - Refactor ldc error codes
    - Correct signature-check
    - Update J9ROMCONSTANTDYNAMICREF_NAMEANDSIGNATURE macro
```

File: runtime/oti/j9nonbuilder.h [hunk 5]
```diff
@@ -2244,8 +2244,7 @@ typedef struct J9ROMFieldRef {
 	J9SRP nameAndSignature;
 } J9ROMFieldRef;
 
-#define J9ROMFIELDREF_NAMEANDSIGNATURE(base) NNSRP_GET((base)->nameAndSignature, struct J9ROMNameAndSignature*)
```
[the remark this answers is not in the document]
Comment (taliamccormick) [reply runtime/oti/j9nonbuilder.h]: Woops, yes - thanks

**Comment by taliamccormick** [comment]
Closing since it is/will be replaced by https://github.com/eclipse/openj9/pull/1858

**Pull request closed without merging**